Added Typescript Typing

PaitoAnderson opened the pull request.

PaitoAnderson committed:
Added Typescript Typing

stevemao reviewed (approved):
I'm not an expert of typescript. But LGTM

stevemao commented:
It looks slightly different than https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/left-pad/index.d.ts
I'll merge this if there's more up votes.

stevemao commented:
@PaitoAnderson any comments on https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/left-pad/index.d.ts?

PaitoAnderson commented:
Thanks for looking this over, I actually didn't notice there was a `@types/left-pad` package or I would have used it.  While both work, I am going to assume theirs is better.

If your cool with it and because it's a much nicer experience to just add one package and have types include I'll update my PR to just use this instead?

Cheers!

stevemao commented:
For the typing, I prefer to keep it one place. But I don't mind duplicate it here for convenience if asked by most users.
Either way, we should add a note in the readme about typing.

PaitoAnderson committed:
Copied Typings for Convenience

Copied from: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/left-pad/index.d.ts

PaitoAnderson committed:
Update README.md

PaitoAnderson commented:
Sounds good. I've done both, let me know if that's good.

stevemao commented:
I believe we need to add `"types": "./index.d.ts",` in package.json

PaitoAnderson committed:
Added types to package.json

PaitoAnderson commented:
You're right, they talk about it here in publishing.

>Also note that if your main declaration file is named index.d.ts and lives at the root of the package (next to index.js) you do not need to mark the "types" property, though it is advisable to do so.

https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html

Crazometer commented:
If this broke your project and you use typescript then you should use

`import leftPad = require('left-pad');`

instead of

`import * as leftPad from 'left-pafd';`

Related:
https://www.typescriptlang.org/docs/handbook/modules.html#export--and-import--require
https://stackoverflow.com/questions/40294870/module-exports-vs-export-default-in-node-js-and-es6

PaitoAnderson commented:
‘import leftPad from 'left-pad';’ works too

timm-gs commented:
@PaitoAnderson: It does not for me:

```
 error TS1192: Module '"node_modules/left-pad/index"' has no default export.
```

I am using TS 2.6.1 with the following config:
```
{
  "include": [
    "packages/*/src/**/*",
    "packages/*/test/**/*"
  ],
  "exclude": [
  ],
  "compilerOptions": {
    /* Basic Options */
    "target": "es2016",  
    "module": "commonjs",   
    "outDir": "./lib",     
    "rootDir": ".",       
    "removeComments": false,
    "strict": true,           
    "noUnusedLocals": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
    "strictFunctionTypes": false, 
    /* Experimental Options */
    "experimentalDecorators": true,   
    "sourceMap": true,
    "noEmitOnError": true,
    "downlevelIteration": true,
    "declaration": true,
    "lib": [
        "esnext",
        "dom"
    ],
    "jsx": "react"
  }
}

```

centigrade-thomas-becker commented:
I have tried several different methods of importing typings for left-pad, but none seem to work properly.

```
// import { leftPad } from 'left-pad';
// import leftPad from 'left-pad';
// import * as leftPad from 'left-pad';

const leftPad = require('left-pad'); // <= the only one that does not produce an error
```

PaitoAnderson commented:
Not sure if it helps but here is the tsconfig.json I use with React Native:

```{
    "compilerOptions": {
  /* Basic Options */
      "target": "es2015",                       /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', or 'ESNEXT'. */
      "module": "es2015",                       /* Specify module code generation: 'commonjs', 'amd', 'system', 'umd' or 'es2015'. */
      // "lib": [],                             /* Specify library files to be included in the compilation:  */
      "allowJs": true,                          /* Allow javascript files to be compiled. */
      // "checkJs": true,                       /* Report errors in .js files. */
      "jsx": "react-native",                    /* Specify JSX code generation: 'preserve', 'react-native', or 'react'. */
      // "declaration": true,                   /* Generates corresponding '.d.ts' file. */
      "sourceMap": true,                        /* Generates corresponding '.map' file. */
      // "outFile": "./",                       /* Concatenate and emit output to single file. */
      // "outDir": "./build",                   /* Redirect output structure to the directory. */
      "rootDir": "./src",                       /* Specify the root directory of input files. Use to control the output directory structure with --outDir. */
      "removeComments": false,                  /* Do not emit comments to output. */
      // "noEmit": true,                        /* Do not emit outputs. */
      // "importHelpers": true,                 /* Import emit helpers from 'tslib'. */
      // "downlevelIteration": true,            /* Provide full support for iterables in 'for-of', spread, and destructuring when targeting 'ES5' or 'ES3'. */
      // "isolatedModules": true,               /* Transpile each file as a separate module (similar to 'ts.transpileModule'). */
  
      /* Strict Type-Checking Options */
      "strict": true,                           /* Enable all strict type-checking options. */
      "noImplicitAny": true,                    /* Raise error on expressions and declarations with an implied 'any' type. */
      "strictNullChecks": true,                 /* Enable strict null checks. */
      "noImplicitThis": true,                   /* Raise error on 'this' expressions with an implied 'any' type. */
      "alwaysStrict": true,                     /* Parse in strict mode and emit "use strict" for each source file. */
  
      /* Additional Checks */
      // "noUnusedLocals": true,                /* Report errors on unused locals. */
      // "noUnusedParameters": true,            /* Report errors on unused parameters. */
      // "noImplicitReturns": true,             /* Report error when not all code paths in function return a value. */
      // "noFallthroughCasesInSwitch": true,    /* Report errors for fallthrough cases in switch statement. */
  
      /* Module Resolution Options */
      "moduleResolution": "node",               /* Specify module resolution strategy: 'node' (Node.js) or 'classic' (TypeScript pre-1.6). */
      // "baseUrl": "./",                       /* Base directory to resolve non-absolute module names. */
      // "paths": {},                           /* A series of entries which re-map imports to lookup locations relative to the 'baseUrl'. */
      // "rootDirs": [],                        /* List of root folders whose combined content represents the structure of the project at runtime. */
      // "typeRoots": [],                       /* List of folders to include type definitions from. */
      // "types": [],                           /* Type declaration files to be included in compilation. */
      "allowSyntheticDefaultImports": true      /* Allow default imports from modules with no default export. This does not affect code emit, just typechecking. */
  
      /* Source Map Options */
      // "sourceRoot": "./",                    /* Specify the location where debugger should locate TypeScript files instead of source locations. */
      // "mapRoot": "./",                       /* Specify the location where debugger should locate map files instead of generated locations. */
      // "inlineSourceMap": true,               /* Emit a single file with source maps instead of having a separate file. */
      // "inlineSources": true,                 /* Emit the source alongside the sourcemaps within a single file; requires '--inlineSourceMap' or '--sourceMap' to be set. */
  
      /* Experimental Options */
      // "experimentalDecorators": true,        /* Enables experimental support for ES7 decorators. */
      // "emitDecoratorMetadata": true,         /* Enables experimental support for emitting type metadata for decorators. */
    }
  }
```

stevemao commented:
```js
import leftPad from 'left-pad';
// or
const leftPad = require('left-pad');
```

Are supposed to work. If not, just send a PR to fix it.

Crazometer commented:
```
import leftPad = require('left-pad');
```

Is the correct syntax. See the links in my comment above.

stevemao commented:
Ok, my bad. Thanks @Crazometer !

tofoli commented:
Change `index.d.ts`
```diff
7c7
< export = leftPad;
---
> export default leftPad;
```
